### ISSUE TITLE
Eliminate use of attributes during config resolving

### ISSUE THREAD

**Issue by geeknoid** [istio/mixer#446]
Right now, we look at attributes to evaluate config selectors. Unless there's a compelling reason, I'd like to eliminate attributes at this level. Instead:

- We need to add a service_name field to all of our API request protos. This will be used to select the  chunk of config to use for the request.

- Finally, the method used is also used to select the chunk of config.

- We need to add an event_name field to all our API request protos. This will also be used to select the chunk of config to use. For the proxy, the event_name would be proxy_ingress and proxy_egress reflecting the direction of traffic. For services calling mixer, there would be a service-defined set of possible event names.

In other words, a request comes in and is classified using service_name:method_name:event_name. This is the top-level selection the mixer performs to decide the set of aspects to invoke.

This change makes it so attribute evaluation is strictly performed within the context of a Check, Report, or Quota method call. This is cleaner and will enable smooth integration with a language-based approach to processing attributes. The operator will be expected to write functions that implement the attribute selection logic for Check, Report, and Quota.

**Comment by wora** [comment]
Totally agree. That is precisely we did with Service Control.

**Comment by mandarjog** [comment]
I am going to eliminate the need for selectors during policy selection as part of my config PR.
It will be based ok key value lookups.

If we take a service producer only view, then it is possible to add a `service` and `service_version` field to the protocol. 

However there are many policies that can and __do__ apply to a single transaction. 
Consider service: `svc1.ns1.cluster.local` 
Just from the service producer perspective there can be 3 policies that apply.
 - service level: Attached to directly to the service. (Modeled by service control). 
 - namespace level: Attached to the namespace
 - cluster level: attached to the cluster

This hierarchy can be arbitrarily deep  (for example you could have `svc1.ns1.cluster1.project`
We cannot add more fields to the protocol for every installation of Istio.

Similarly, all this is just from the service producer perspective. There can be consumer perspective, network owner perspective and many others. 

I have been calling these __identity__ attributes, because they help identify traffic. 
So I agree that these are not normal payload attributes, but we need to elevate them to protocol in an extensible way, or they leave them as is. The set of identity attributes is a property of the system.

**Comment by geeknoid** [comment]
You raise some good points. 

I actually don't object to having the state carried as attributes, and up using attributes to do the high-level selection within config. It does provide ultimate flexibility. What I do have an issue with is the need for expression evaluation of these things, since it will break the language model. If we can process these identity attributes entirely without needing expressions, then I'm happy.

If we do need the expressiveness of expressions for this stuff, then it means the language option we've been looking at needs to be more holistic. Whereas I had been thinking of a number of distinct functions, one per service:method:case, instead we'd only separate incoming traffic by method and dispatch to one of three separate functions. The functions would then explicitly look at attributes to decide what to do:

    if attribute.service == "myService" {
        // do the service 1 stuff
    } else ...

And so forth.

Basically, this ends up replacing the adapter manager and aspect managers completely, replaced by custom code written specifically to understand the set of attributes and descriptors in the system.

**Comment by douglas-reid** [comment]
Does this really need to be for alpha? This language model definitely seems like something we should be investigating post-alpha.

I'm a bit wary of assuming a nebulous language model and then designing around it. We haven't even finished the alpha version of config yet.

**Comment by mandarjog** [comment]
pr #525 fixes this.
